### PR TITLE
WPCOM Plugins: use software slug instead of slug

### LINKED ISSUE #67535: Marketplace: Slug collision handling
A plugin slug collision is when a dotcom or woocommerce.com marketplace plugin has the same slug as one on dotorg. 

It could be different plugins entirely or the same plugin owned by the same developer. (e.g. gravityforms)

Context p1662540187410729/1662499292.053829-slack-CKZHG0QCR

In this issue we only need to be concerned with the unexpected collision problem, e.g. someone adds a new dotorg plugin with an existing dotcom slug.

To avoid the dotorg slug taking over our dotcom listing, we can make sure dotcom overrides dotorg [here](https://github.com/Automattic/wp-calypso/blob/8d0fe3f90dc85a490ea7334f8df73d0d8f1e694b/client/my-sites/plugins/plugin-details.jsx#L168-L182).

- [ ] Override dotorg plugins with dotcom plugins whenever looking up plugins by slug

### PR DESCRIPTION
**Pull request by cpapazoglou**
#### Proposed Changes

* Uses the `software_slug` to identify if a premium plugin is installed
* checks against the `software_slug` and plugin purchase to show whether plugin is installed "correctly"

#### Testing Instructions

* In a site with premium plugins purchased verify that they correctly state installed and active status
* Visit `http://calypso.localhost:3000/plugins/browse/paid/visualcomposertesting.wpcomstaging.com` and `http://calypso.localhost:3000/plugins/visualcomposer-wpcom/visualcomposertesting.wpcomstaging.com` and verify that visual composer correctly displays as installed and active.

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Fixes #67535

### CODE REVIEW

File: client/lib/plugins/utils.js [hunk 1]
```diff
@@ -104,6 +104,7 @@ export function getAllowedPluginData( plugin ) {
 		'name',
 		'network',
 		'num_ratings',
+		'org_slug',
```
Comment (DustyReagan) [inline client/lib/plugins/utils.js]: What do we use 'org_slug' for? I noticed it hasn't really been used much (or maybe ever?). Do we need it?

Comment (cpapazoglou) [reply client/lib/plugins/utils.js]: `org_slug` is a chance for us to match a premium plugin to its .org counterpart and thus show an upsell or something similar in the .org plugin details page. Related to https://github.com/Automattic/wp-calypso/issues/60769

Comment (DustyReagan) [reply client/lib/plugins/utils.js]: Got it! Thanks for the context!

File: client/lib/plugins/utils.js [hunk 2]
```diff
@@ -424,3 +426,25 @@ export function getProductSlugByPeriodVariation( periodVariation, productsList )
...
+/**
+ * @param  {object} plugin The plugin object
+ * @param  {Array} purchases An array of site purchases
+ * @param  {boolean} isMarketplaceProduct Is this part of WP.com Marketplace or WP.org
+ * @returns {string} The software slug string
+ */
+export const isPluginPurchased = ( plugin, purchases, isMarketplaceProduct ) =>
+	isMarketplaceProduct &&
+	purchases.find( ( purchase ) =>
+		Object.values( plugin?.variations ).some(
```
Comment (DustyReagan) [inline client/lib/plugins/utils.js]: Can `plugin?.variations` be undefined sometimes? On http://calypso.localhost:3000/plugins/browse/paid/example.wpcomstaging.com, sometimes when I click on a plugin to go to the plugin details page, I get a "We're sorry, but an unexpected error has occurred" error page. It's a bit inconsistent (not always the same plugin), but stepping through the browser debugger, it looks like `plugin?.variations` is sometimes undefined?


https://user-images.githubusercontent.com/140841/195126006-e57262ad-0e41-4eac-80f4-e0511a6d91e3.mp4

**Comment by DustyReagan** [comment]
On the plugin details page of Visual Composer, when I click on "Deactivate" I get a console error `Uncaught TypeError: can't access property "active", plugin is undefined`.

**Comment by cpapazoglou** [comment]
> On the plugin details page of Visual Composer, when I click on "Deactivate" I get a console error `Uncaught TypeError: can't access property "active", plugin is undefined`.

Good catch, handled in [708ba83](https://github.com/Automattic/wp-calypso/pull/68602/commits/708ba833cbc3f0b8830516c912e2fb58ba43e6cc)

File: client/lib/plugins/utils.js (hunk 2 above)
Comment (cpapazoglou) [reply client/lib/plugins/utils.js]: This should be addressed here https://github.com/Automattic/wp-calypso/pull/68602/commits/2c15ccf1f35b91c65fd96639cf5ba17f5e1f6489

Comment (DustyReagan) [reply client/lib/plugins/utils.js]: I'm still seeing the same behavior. Maybe `getPluginPurchased` is trying to return `undefined`?

**Comment by DustyReagan** [comment]
> >On the plugin details page of Visual Composer, when I click on "Deactivate" I get a console error Uncaught TypeError: can't access property "active", plugin is undefined.
> 
> Good catch, handled in [708ba83](https://github.com/Automattic/wp-calypso/pull/68602/commits/708ba833cbc3f0b8830516c912e2fb58ba43e6cc)

Nice! Activate and Deactivate seem to be working. I noticed that on other plugins there is this feedback message onclick. It doesn't seem to appear on Visual Composer.

File: client/lib/plugins/utils.js (hunk 2 above)
Comment (cpapazoglou) [reply client/lib/plugins/utils.js]: Caught the error 

CleanShot 2022-10-14 at 18 03 25@2x

I tried to reproduce it afterwards without luck. Seems being a concurrency issue with the plugin being `null` or `undefined` while being used in `Object.values( plugin.variations )`. This should fix that [c840220](https://github.com/Automattic/wp-calypso/pull/68602/commits/c840220951453e3236964be9db1f2a63e1bd7603). Can you test again?

**Comment by cpapazoglou** [comment]
> > > On the plugin details page of Visual Composer, when I click on "Deactivate" I get a console error Uncaught TypeError: can't access property "active", plugin is undefined.
> > 
> > 
> > Good catch, handled in [708ba83](https://github.com/Automattic/wp-calypso/pull/68602/commits/708ba833cbc3f0b8830516c912e2fb58ba43e6cc)
> 
> Nice! Activate and Deactivate seem to be working. I noticed that on other plugins there is this feedback message onclick. It doesn't seem to appear on Visual Composer.
> 
> 

Good catch. Since it doesn't seem blocking I would like us to follow up in a separate issue. What do you think?

**Comment by DustyReagan** [comment]
> Good catch. Since it doesn't seem blocking I would like us to follow up in a separate issue. What do you think?

Sounds good to me!

File: client/lib/plugins/utils.js (hunk 2 above)
Comment (DustyReagan) [reply client/lib/plugins/utils.js]: That seems to have done the trick! It's working well for me now! 👍

**Review by DustyReagan — APPROVED** [review]
Working great! 👍 

Side note, the unit tests are failing, but I don't think it's related to this PR. 🤔 I tried rebasing but it didn't seem to help.

**Review by paulopmt1 — APPROVED** [review]
Good job @cpapazoglou! It worked great at Have you tested the feature in Simple, Atomic, and self-hosted Jetpack sites. 

I noted the same issue as Dusty and agree it could be a follow-up task because it only occurs with `visual-composer-dotcom`

**Pull request merged**